File: src/main/java/com/github/zyypj/tadeuBooter/api/collections/GenericCache.java
package com.github.zyypj.tadeuBooter.api.collections;

import lombok.AllArgsConstructor;
import lombok.Data;

import javax.annotation.Nullable;
import java.util.*;
import java.util.function.Supplier;

@Data
@AllArgsConstructor
public abstract class GenericCache<K, V, T> {

    protected final Map<K, V> cache;

    protected GenericCache() {
        cache = getInitialMap();
    }

    public void add(T t) {
        MappablePair<K, V> apply = apply(t);
        cache.put(apply.getFirst(), apply.getSecond());
    }

    public void addAll(Collection<T> collection) {
        for (T t : collection) add(t);
    }

    public void addAll(Map<K, V> map) {
        cache.putAll(map);
    }

    public Optional<V> remove(@Nullable K key) {
        if (key == null) return Optional.empty();
        return Optional.ofNullable(cache.remove(key));
    }

    public boolean has(@Nullable K key) {
        return key != null && cache.containsKey(key);
    }

    public Optional<V> get(@Nullable K key) {
        if (key == null) return Optional.empty();
        return Optional.ofNullable(cache.get(key));
    }

    public V getOrInsert(K key, T def) {
        if (cache.containsKey(key)) return cache.get(key);
        MappablePair<K, V> pair = apply(def);
        add(def);
        return pair.getSecond();
    }

    public V getOrInsert(K key, Supplier<T> defaultValue) {
        if (cache.containsKey(key)) return cache.get(key);
        T def = defaultValue.get();
        if (def == null) throw new NullPointerException("The supplier returned a null value");
        MappablePair<K, V> pair = apply(def);
        add(def);
        return pair.getSecond();
    }

    public Set<K> keySet() {
        return cache.keySet();
    }

    public Collection<V> getAll() {
        return cache.values();
    }

    public void clear() {
        cache.clear();
    }

    public int size() {
        return cache.size();
    }

    public boolean isEmpty() {
        return cache.isEmpty();
    }

    public abstract MappablePair<K, V> apply(T t);

    protected Map<K, V> getInitialMap() {
        return new HashMap<>();
    }

    public List<V> get(Iterable<K> keys) {
        List<V> result = new ArrayList<>(size());
        for (K key : keys) get(key).ifPresent(result::add);
        return result;
    }
}

File: src/main/java/com/github/zyypj/tadeuBooter/api/collections/GenericCache.kt
package com.github.zyypj.tadeuBooter.api.collections

abstract class GenericCache<K, V, T> {
    protected val cache: MutableMap<K, V> = getInitialMap()

    protected open fun getInitialMap(): MutableMap<K, V> = mutableMapOf()

    abstract fun apply(t: T): Pair<K, V>

    fun add(t: T) {
        val (k, v) = apply(t)
        cache[k] = v
    }

    fun addAll(collection: Collection<T>) = collection.forEach(::add)

    fun addAll(map: Map<K, V>) = cache.putAll(map)

    fun remove(key: K?): V? = key?.let(cache::remove)

    fun has(key: K?): Boolean = key != null && cache.containsKey(key)

    fun get(key: K?): V? = key?.let(cache::get)

    fun getOrInsert(key: K, def: T): V = cache[key] ?: run {
        val (k, v) = apply(def)
        cache[k] = v
        v
    }

    fun getOrInsert(key: K, defaultValue: () -> T): V = cache[key] ?: run {
        val def = defaultValue() ?: throw NullPointerException("Supplier retornou null")
        val (k, v) = apply(def)
        cache[k] = v
        v
    }

    fun keySet(): Set<K> = cache.keys

    fun getAll(): Collection<V> = cache.values

    fun clear() = cache.clear()

    fun size(): Int = cache.size

    fun isEmpty(): Boolean = cache.isEmpty()

    fun get(keys: Iterable<K>): List<V> = keys.mapNotNull { cache[it] }
}